File: 42_Trapping Rain Water.java
// Time complexity: O(n)
// Space complexity: O(1)

// trap water based on the lower wall
// 1. calculate trapped water
// 2. update pointer
// 3. update wall height

class Solution {
    public int trap(int[] height) {
        int res = 0, l = 0, r = height.length - 1, lMax = height[0], rMax = height[height.length-1];

        while (l < r) {
            lMax = Math.max(lMax, height[l]);
            rMax = Math.max(rMax, height[r]);
            if (lMax < rMax) {
                res += lMax - height[l];
                l += 1;
            }
            else {
                res += rMax - height[r];
                r -= 1;
            }
        }
        return res;
    }
}

File: 42_Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        # 2-Pointer: shift towards center
        # the lower boundary determines the trapped amount
        res, l, r, lMax, rMax = 0, 0, len(height) - 1, 0, 0

        while l <= r:
            if lMax < rMax:
                lMax = max(lMax, height[l])
                res += lMax - height[l]
                l += 1
            else:
                rMax = max(rMax, height[r])
                res += rMax - height[r]
                r -= 1

        return res